processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,344 @@
+
+	// traceID to be removed
+	removed chan []byte
+
+	// shutdown synchronization
+	stopped        chan context.Context // used in the main loop
+	cancelReleases bool                 // used for traces that are just waiting for the duration
+	cancelLock     *sync.RWMutex
+}
+
+var _ component.TraceProcessor = (*groupByTraceProcessor)(nil)
+
+// newGroupByTraceProcessor returns a new processor.
+func newGroupByTraceProcessor(logger *zap.Logger, st storage, nextConsumer consumer.TraceConsumer, config Config) (*groupByTraceProcessor, error) {
+	if nextConsumer == nil {
+		return nil, componenterror.ErrNilNextConsumer
+	}
+
+	sp := &groupByTraceProcessor{
+		logger:       logger.With(zap.String("processor", "groupbytrace")),
+		nextConsumer: nextConsumer,
+		config:       config,
+		st:           st,
+		ids:          make([][]byte, config.NumTraces),
+		idToIndex:    make(map[uint64]int),
+		index:        -1, // the first span to be received will be placed at position '0'
+
+		// open the channels
+		removed:  make(chan []byte, 50), // TODO: perf tests to decide what's a good buffer size
+		expired:  make(chan []byte, 50),
+		received: make(chan pdata.Traces, 50),
+		released: make(chan *v1.ResourceSpans, 50),
+
+		// shutdown synchronization
+		stopped:        make(chan context.Context),
+		cancelLock:     &sync.RWMutex{},
+		cancelReleases: false,
+	}
+
+	return sp, nil
+}
+
+func (sp *groupByTraceProcessor) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+	sp.received <- td
+	return nil
+}
+
+func (sp *groupByTraceProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (sp *groupByTraceProcessor) Start(_ context.Context, _ component.Host) error {
+	// starts the processes listening to the channels
+	go sp.onReleased()
+	go sp.onRemoved()
+	go sp.onTrace()
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (sp *groupByTraceProcessor) Shutdown(ctx context.Context) error {
+	sp.stopped <- ctx
+	return nil
+}
+
+func (sp *groupByTraceProcessor) onTrace() {
+	for {
+		// We have a "main loop" here, to avoid real locks in the hot path. The only lock we have
+		// is in the shutdown path, and the processor's state is checked in the funcs that run
+		// in the future.
+		// The other two buffered channels do modifications to the internal map/ring as part of the same thread
+		// as the main loop, so, there's only one access at any given time. Potentially blocking operations
+		// should be done asynchronously, which isn't 100% the case in this initial implementation.
+		select {
+		case ctx := <-sp.stopped:
+			sp.processShutdown(ctx)
+			return
+		case traceID := <-sp.expired:
+			// in this implementation, this won't return errors
+			sp.processExpired(traceID)
+		case item := <-sp.received:
+			// in this implementation, this won't return errors
+			batch := pdata.TracesToOtlp(item)
```

Comment:
Why do you need to convert to OTLP model rather than iterating through `item.ResourceSpans()`?

Reply:
I'll double check, but I think I needed something from the protobuf models. It might have something to do with the trace ID.